### PR TITLE
[SPARK-17051][SQL] we should use hadoopConf in InsertIntoHiveTable

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Hive confs in hive-site.xml will be loaded in `hadoopConf`, so we should use `hadoopConf` in `InsertIntoHiveTable` instead of `SessionState.conf`
## How was this patch tested?

N/A

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @yhuai

**Comment by gatorsmile** [comment]
Great! It resolves my original concern. Thanks!

**Comment by gatorsmile** [comment]
`hive.exec.dynamic.partition` also impacts our regular writing paths (DataFrameWriter APIs). I remember DataFrameWriter APIs always assume this conf is true, right?

If it is controlled by hadoopConf, users might hit strange errors when using our DataFrameWriter APIs.

**Comment by cloud-fan** [comment]
> hive.exec.dynamic.partition also impacts our regular writing paths

I think it's hive only conf? Normal data source relation should not read this conf.

**Comment by gatorsmile** [comment]
If users want to use the `DataFrameWriter`'s `insertInto` API for partitioned Hive table, they have to set `hive.exec.dynamic.partition` to `true` and `hive.exec.dynamic.partition.mode` to `nonstrict`. Otherwise, it does not work. Users are unable to specify the partition values in the `insertInto` APIs. See [the code](https://github.com/apache/spark/blob/8c8acdec9365136cba13060ce36c22b28e29b59b/sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala#L258). 

Let me use a test case to show the issue. 
```Scala
    withTempDir { tmpDir =>
      val basePath = tmpDir.getCanonicalPath
      val externalTab = "extTable_with_partitions"
      withTable(externalTab) {
        assert(tmpDir.listFiles.isEmpty)
        sql(
          s"""
             |CREATE EXTERNAL TABLE $externalTab (key INT, value STRING)
             |PARTITIONED BY (ds STRING, hr STRING)
             |stored as Parquet
             |LOCATION '$basePath'
           """.stripMargin)

        for (ds <- Seq("2008-04-08", "2008-04-09"); hr <- Seq("11", "12")) {
          sql(
            s"""
               |INSERT OVERWRITE TABLE $externalTab
               |partition (ds='$ds',hr='$hr')
               |SELECT 1, 'a'
             """.stripMargin)
        }
        withSQLConf("hive.exec.dynamic.partition" -> "true",
            "hive.exec.dynamic.partition.mode" -> "nonstrict") {
          Seq((1, "2", "2008-04-09", "12")).toDF("key", "value", "ds", "hr").write
            .insertInto(externalTab)
        }
      }
    }
```

**Comment by gatorsmile** [comment]
Sorry, let me rephrase the potential issue. 

- `insertInto` API forces users to set `hive.exec.dynamic.partition` to `true` and `hive.exec.dynamic.partition.mode` to `nonstrict`. This might not be convinient. The default value of `hive.exec.dynamic.partition.mode` is `strict`. 

- If we always read the setting from `hadoopConf`, does that mean users are unable to control these settings for different queries? Is that possible users can change the setting values in `hadoopConf` at runtime?

**Comment by cloud-fan** [comment]
1. `hadoopConf` contains all the confs from SQL conf, see `SessionState.newHadoopConf`
2. users can change `hadoopConf` at runtime.
3. the default value is from hive, I'm ok to not follow it.

**Comment by gatorsmile** [comment]
uh, I see. Thank you! No more question. : )

**Comment by yhuai** [comment]
Sorry. What's the necessity to make this change?

**Comment by gatorsmile** [comment]
Based on my understanding, after this PR, we will respect the conf values of `hive.exec.dynamic.partition`, `hive.exec.dynamic.partition.mode` and `hive.exec.compress.output` that are specified in `hive-site.xml`?

**Comment by JoshRosen** [comment]
@yhuai, @cloud-fan What's the status of this issue? Should this still be targeted for 2.0.1 (which it is currently in JIRA)?

**Comment by cloud-fan** [comment]
cc @yhuai , In `InsertIntoHiveTable` we already called `newHadoopConf`, I think it's safer to get these hive confs from `hadoopConf` instead  of `sqlConf`, to respect hive confs in `hive-site.xml`. We can check other places one by one later.

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
@JoshRosen this is a regression in 2.0(it works in 1.6), so I think we should target it to 2.0

**Comment by yhuai** [comment]
This change looks good. Let's add a regression test.

**Comment by yhuai** [comment]
LGTM. Merging to master and branch 2.0.
